finality-provider/service/finality_submitter.go

```diff
@@ -0,0 +1,401 @@
+}
+
+func (ds *DefaultFinalitySubmitter) getBtcPk() *btcec.PublicKey {
+	return ds.state.GetBtcPk()
+}
+
+func (ds *DefaultFinalitySubmitter) getBtcPkBIP340() *bbntypes.BIP340PubKey {
+	return ds.state.GetBtcPkBIP340()
+}
+
+func (ds *DefaultFinalitySubmitter) mustSetLastVotedHeight(height uint64) {
+	if err := ds.state.SetLastVotedHeight(height); err != nil {
+		ds.logger.Fatal("failed to update state after finality signature submitted",
+			zap.String("pk", ds.getBtcPkHex()), zap.Uint64("height", height), zap.Error(err))
+	}
+}
+
+func (ds *DefaultFinalitySubmitter) mustSetStatus(s proto.FinalityProviderStatus) {
+	if err := ds.state.SetStatus(s); err != nil {
+		ds.logger.Fatal("failed to set finality-provider status",
+			zap.String("pk", ds.getBtcPkHex()), zap.String("status", s.String()))
+	}
+}
+
+// SetState sets the finality provider state store.
+func (ds *DefaultFinalitySubmitter) SetState(state types.FinalityProviderState) {
+	ds.state = state
+}
+
+// FilterBlocksForVoting filters blocks based on the finality provider's voting power and height criteria for submission.
+// It returns a slice of blocks eligible for voting and an error if any issues are encountered during processing.
+// It also updates the fp instance status according to the block's voting power
+func (ds *DefaultFinalitySubmitter) FilterBlocksForVoting(ctx context.Context, blocks []types.BlockDescription) ([]types.BlockDescription, error) {
+	processedBlocks := make([]types.BlockDescription, 0, len(blocks))
+
+	var hasPower bool
+	var err error
+	for _, b := range blocks {
+		blk := b
+		if blk.GetHeight() <= ds.state.GetLastVotedHeight() {
+			ds.logger.Debug(
+				"the block height is lower than last processed height",
+				zap.String("pk", ds.getBtcPkHex()),
+				zap.Uint64("block_height", blk.GetHeight()),
+				zap.Uint64("last_voted_height", ds.state.GetLastVotedHeight()),
+			)
+
+			continue
+		}
+
+		// check whether the finality provider has voting power
+		hasPower, err = ds.getVotingPowerWithRetry(ctx, blk.GetHeight())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get voting power for height %d: %w", blk.GetHeight(), err)
+		}
+		if !hasPower {
+			ds.logger.Debug(
+				"the finality-provider does not have voting power",
+				zap.String("pk", ds.getBtcPkHex()),
+				zap.Uint64("block_height", blk.GetHeight()),
+			)
+
+			// the finality provider does not have voting power
+			// and it will never will at this block, so continue
+			ds.metrics.IncrementFpTotalBlocksWithoutVotingPower(ds.getBtcPkHex())
+
+			continue
+		}
+
+		processedBlocks = append(processedBlocks, blk)
+	}
+
+	// update fp status according to the power for the last block
+	if hasPower && ds.state.GetStatus() != proto.FinalityProviderStatus_ACTIVE {
+		ds.mustSetStatus(proto.FinalityProviderStatus_ACTIVE)
+	}
+
+	if !hasPower && ds.state.GetStatus() == proto.FinalityProviderStatus_ACTIVE {
+		ds.mustSetStatus(proto.FinalityProviderStatus_INACTIVE)
+	}
```

Comment:
I'm not sure if this function is supposed to modify FP's state in DB. This operation seems to belong to certain goroutines. Wdyt?